bleep: init at 0.0.7

KristianAN opened the pull request:
## Description of changes

Added new package: Bleep - A fast modern build tool for Scala with a declarative configuration: https://bleep.build

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [X] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [X] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Add a 👍 [reaction](https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to [issues you find important](https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

Commit:
maintainers: add KristianAN

KristianAN commented:
Is there anything missing for this to be merged?

NobbZ commented:
As you merged `master` back into your PR, getting the commits right now might become a mess easily, therefore feel free to close and recreate the PR, making sure that you follow the [commit conventions](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions).

So once you got the commits right, either by rebase/squash/fixup on your current PR or by creating a fresh one, feel free to ping me for a proper review.

----

Some additional notes:

1. Please move `platform` and `hash` into a `let`-binding. Especially `platform` doesn't seem to be used for anything but constructing the URL.
2. Instead of using `is*` predicates use "poor mans switch" to construct platform and hash, eg.
   ```nix
   let platform = {
     x86_64-linux = "x86_64-pc-linux";
     aarch64-darwin = "arm64-apple-darwin";
     # etc
   }."${stdenv.system}"; in …
   ```

KristianAN commented:
@NobbZ I think everything should be in order now. The platform names I assume should be the ones defined in lib.platform.<target>, i.e. libs.platform.aarch64 which gives aarch64-darwin.

NobbZ reviewed (changes requested):
Please use distinct commits for adding yourself as a maintainer and adding the package.

1. `maintainers: add KristianAN`
2. `bleep: init at 0.0.7`

Commit:
bleep: init at 0.0.7

Scala build tool bleep.
https://github.com/oyvindberg/bleep/releases/tag/v0.0.7
A fast modern build tool for Scala with a declarative configuration.

Pandapip1 commented on pkgs/by-name/bl/bleep/package.nix:
```diff
@@ -0,0 +1,89 @@
+    }
+    ."${stdenv.system}";
+in
+stdenv.mkDerivation rec {
```
Use the `finalAttrs` pattern instead of `rec`: https://nixos.org/manual/nixpkgs/stable/#mkderivation-recursive-attributes

Pandapip1 commented on pkgs/by-name/bl/bleep/package.nix:
```diff
@@ -0,0 +1,89 @@
+    license = licenses.mit;
+    description = "Bleeping fast scala build tool";
+    mainProgram = "bleep";
+    platforms = [
+      "x86_64-linux"
+      "x86_64-apple-darwin"
+      "arm64-apple-darwin"
+    ];
```
These platform strings look wrong for darwin.

Pandapip1 commented on pkgs/by-name/bl/bleep/package.nix:
```diff
@@ -0,0 +1,89 @@
+        --zsh <(bleep install-tab-completions-zsh --stdout) \
+    '';
+
+  meta = with lib; {
```
`meta = with lib; ` is considered an anti-pattern. Remove the `with lib;`.

Pandapip1 commented on pkgs/by-name/bl/bleep/package.nix:
```diff
@@ -0,0 +1,89 @@
+  pname = "bleep";
+  version = "0.0.7";
+
+  src = fetchurl {
+    url = "https://github.com/oyvindberg/bleep/releases/download/v${version}/bleep-${platform}.tar.gz";
+    sha256 = hash;
+  };
```
Why aren't you using `fetchFromGitHub` and building from source? And if you're going to use prebuilt binaries, you should use `stdenvNoCC`.

KristianAN replied:
I changed to NoCC. Bleep builds bleep itself (including graalvm-native-image packaging) and I have no idea on how I would bootstrap this.

Pandapip1 replied:
> Bleep builds bleep itself

That seems... flawed.

KristianAN commented:
@NobbZ is this good to merge?

Pandapip1 commented on pkgs/by-name/bl/bleep/package.nix:
```diff
@@ -0,0 +1,91 @@
+    stdenv.cc.cc
+  ];
+
+  unpackPhase = ''
+    runHook preUnpack
+    tar -xf $src
+    runHook postUnpack
+  '';
```
Forgive me if I'm wrong, but isn't this `unpackPhase` redundant? I thought `.tar.gz` were automatically unpacked.

d-brasher replied:
I think that is true if you use [`fetchzip`](https://nixos.org/manual/nixpkgs/stable/#sec-pkgs-fetchers-fetchzip) (which seems like the most appropriate fetcher to use).

NobbZ replied:
`fetchzip` does unpack as part of its own thing.

`mkDerivation`s default `unpackPhase` does unpack `$src` if it is a tarball or zip (IIRC for ZIP you need to have zip in the `nativeBuildInputs`) or it copies recursively if `$src` is plain folder.

d-brasher replied:
Oh you are right. Which of the two would be the preferred approach (if there is one)?

Pandapip1 commented on pkgs/by-name/bl/bleep/package.nix:
```diff
@@ -0,0 +1,91 @@
+  installPhase = ''
+    runHook preInstall
+    install -Dm755 bleep $out/bin/.bleep-wrapped
+    makeWrapper $out/bin/.bleep-wrapped $out/bin/bleep \
+      --argv0 "$out/bin/bleep"
```
What is the point of this `makeWrapper` call?

d-brasher commented on pkgs/by-name/bl/bleep/package.nix:
```diff
@@ -0,0 +1,91 @@
+    license = lib.licenses.mit;
+    description = "Bleeping fast scala build tool";
+    mainProgram = "bleep";
+    platforms = [
+      "x86_64-linux"
+      "x86_64-darwin"
+      "aarch64-darwin"
+    ];
```
You should probably directly reference the ones in https://github.com/NixOS/nixpkgs/blob/master/lib/systems/doubles.nix

```suggestion
    platforms = with lib.platforms [
      x86_64-linux
      x86_64-darwin
      aarch64-darwin
    ];
```

KristianAN replied:
I tried this, but it only gives me undefined variable. I'm not sure how I am supposed to refer to them?

d-brasher replied:
You are right, disregard :sweat_smile:

SuperSandro2000 commented on pkgs/by-name/bl/bleep/package.nix:
```diff
@@ -0,0 +1,91 @@
+      x86_64-darwin = "x86_64-apple-darwin";
+      aarch64-darwin = "arm64-apple-darwin";
+    }
+    ."${stdenv.system}";
+
+  hash =
+    {
+      x86_64-linux = "sha256-ttfmTMSK6u2zX87CUYjibr675Xzn4/byWub67e5Ir+I=";
+      x86_64-darwin = "sha256-bh1yQ7KZ+cwVHTLG+gpYDLPKG2VUcrz88Sqd2tRiKfM=";
+      aarch64-darwin = "sha256-IV2JPXibv80Odb+1Zkw1YUoutHsi5YaIjyNpAOIUlr4=";
+    }
+    ."${stdenv.system}";
```
Those should get an ``or (throw "unsupported system ${stdenv.hostPlatform.system}")``

KristianAN replied:
Done

SuperSandro2000 commented on pkgs/by-name/bl/bleep/package.nix:
```diff
@@ -0,0 +1,91 @@
+      autoPatchelf $out
+    ''
+    + ''
+      mkdir temp
+      cp $out/bin/.bleep-wrapped temp/bleep
+      PATH="./temp:$PATH"
+
+      installShellCompletion --cmd bleep \
```
```suggestion
      installShellCompletion --cmd $out/bin/bleep \
```

KristianAN replied:
This actually does not work, as it just says command bleep not found.

NobbZ commented on pkgs/by-name/bl/bleep/package.nix:
```diff
@@ -0,0 +1,79 @@
+      aarch64-darwin = "sha256-V8QGF3Dpuy9I6CqKsJRHBHRdaLhc4XKZkv/rI7zs+qQ=";
+    }
+    ."${stdenvNoCC.system}" or (throw "unsupported system ${stdenvNoCC.hostPlatform.system}");
+  version = "0.0.7";
```
```suggestion
```

NobbZ commented on pkgs/by-name/bl/bleep/package.nix:
```diff
@@ -0,0 +1,79 @@
+in
+stdenvNoCC.mkDerivation (finalAttrs: {
+  pname = "bleep";
+  version = version;
```
```suggestion
  version = "0.0.7";
```

NobbZ commented on pkgs/by-name/bl/bleep/package.nix:
```diff
@@ -0,0 +1,79 @@
+  version = version;
+
+  src = fetchzip {
+    url = "https://github.com/oyvindberg/bleep/releases/download/v${version}/bleep-${platform}.tar.gz";
```
```suggestion
    url = "https://github.com/oyvindberg/bleep/releases/download/v${finalAttrs.version}/bleep-${platform}.tar.gz";
```

NobbZ commented on pkgs/by-name/bl/bleep/package.nix:
```diff
@@ -0,0 +1,79 @@
+    runHook postInstall
+  '';
+
+  dontAutoPatchelf = true;
+
+  postFixup =
+    lib.optionalString stdenvNoCC.isLinux ''
+      autoPatchelf $out
+    ''
```
Why are you disabling and then manually autopatching?

d-brasher replied:
It seems that the ELF patching is done _after_ the `postFixup` hook for some reason, and the patched executable is needed for `installShellCompletion`.

d-brasher replied:
It seems that the ELF patching is done _after_ the `postFixup` hook for some reason, and the patched executable is needed for `installShellCompletion`.

NobbZ replied:
Oh, right, makes sense. So this can be considered solved

d-brasher replied:
I still don't understand _why_ it's this way (isn't `patchElfHook` supposed to be part of `fixupPhase` and executed _before_ `postFixup` hook?), but this workaround works I guess :sweat_smile:

NobbZ commented on pkgs/by-name/bl/bleep/package.nix:
```diff
@@ -0,0 +1,79 @@
+      autoPatchelf $out
+    ''
+    + ''
+      mkdir temp
+      cp $out/bin/bleep temp/bleep
+      PATH="./temp:$PATH"
+
+      installShellCompletion --cmd bleep \
+        --bash <(bleep install-tab-completions-bash --stdout) \
+        --zsh <(bleep install-tab-completions-zsh --stdout) \
```
```suggestion
      installShellCompletion --cmd bleep \
        --bash <($out/bin/bleep install-tab-completions-bash --stdout) \
        --zsh <($out/bin/bleep install-tab-completions-zsh --stdout) \
```

d-brasher replied:
Also needs `bleep` in `$PATH`

```suggestion
      export PATH=$PATH:$out/bin
      installShellCompletion --cmd bleep \
        --bash <(bleep install-tab-completions-bash --stdout) \
        --zsh <(bleep install-tab-completions-zsh --stdout) \
```

d-brasher replied:
Also needs `bleep` in `$PATH`

```suggestion
      export PATH=$PATH:$out/bin
      installShellCompletion --cmd bleep \
        --bash <(bleep install-tab-completions-bash --stdout) \
        --zsh <(bleep install-tab-completions-zsh --stdout) \
```

d-brasher commented on pkgs/by-name/bl/bleep/package.nix:
```diff
@@ -0,0 +1,79 @@
+        --bash <(bleep install-tab-completions-bash --stdout) \
+        --zsh <(bleep install-tab-completions-zsh --stdout) \
+    '';
+
```
```suggestion

  passthru.updateScript = nix-update-script { };
  
```

If you want automatic update PRs.

NobbZ replied:
I am opposed to this one, or is it able to cover the fact that we have to maintain 3 hashes for this package?

d-brasher commented:
Result of `nixpkgs-review pr 321599` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>bleep</li>
  </ul>
</details>

d-brasher commented on pkgs/by-name/bl/bleep/package.nix:
```diff
@@ -0,0 +1,75 @@
+        --bash <(bleep install-tab-completions-bash --stdout) \
+        --zsh <(bleep install-tab-completions-zsh --stdout) \
+    '';
+
```
```suggestion

passthru.tests.version = testers.testVersion { package = finalAttrs.finalPackage; };
    
```
If you want you could add something like this as a (very basic) test. (just remember to also add `testers` on top)

d-brasher reviewed (approved):
LGTM :+1:

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1952
